lib/VMwareWebService/broker_timeout.rb

```diff
@@ -0,0 +1,98 @@
+# contains an unprotected critical section that may result in unexpected
+# and/or undesired behavior - depending on the client's usage.
+# In MIQ, this bug contributed to the issue described here:
+#     https://bugzilla.redhat.com/show_bug.cgi?id=1207018
+#
+# Description: The Timeout.timeout method starts a "timing" thread that
+# sleeps for the specified period of time. Once the timing thread is
+# started, the parent (timed) thread yields control to the code block - the
+# section of code subject to the timeout.
+#
+# A timeout occurs when the timing thread wakes up before we return from the 
+# yield to the code block. Here, the timing thread raises an exception in the
+# timed thread, terminating the execution of the code block.
+#
+# When the code block doesn't timeout, we return from the yield. The parent
+# thread then kills the timing thread - supposedly preventing it from waking
+# up and raising a timeout exception. I say supposedly, because this is the
+# heart of the unprotected critical section. From the time the yield returns,
+# until the time the timing thread is killed, the timing thread can still
+# wake up and raise an exception in the timed thread. Since the timed thread
+# is no longer executing the code block subject to timeout, this can result
+# in unpredicted and undesired behavior.
+#
+# This fix employs a Mutex and state flag to protect this critical section
+# of code, preventing this from happening.
+#
+# TODO: This code is based on the Ruby 2.0 implementation of timeout. When
+# upgrading to new versions of Ruby, we need to reevaluate this code against
+# the new Ruby implementations.
+#
+module Timeout
+  def timeout(sec, klass = nil)   #:yield: +sec+
+    return yield(sec) if sec == nil or sec.zero?
+    exception = klass || Class.new(ExitException)
+    state_lock = Mutex.new
+    state = :sleeping
+
+    begin
+      begin
+        x = Thread.current
+        y = Thread.start {
+          begin
+            sleep sec
+          rescue => e
+            state_lock.synchronize do
+              if state == :sleeping
+                state = :exception
+                x.raise e
+              end
+            end
+          else
+            state_lock.synchronize do
+              if state == :sleeping
+                state = :timeout
+                x.raise exception, "execution expired"
+              end
+            end
+          end
+        }
+        rv = yield(sec)
+        state_lock.synchronize do
+          if state == :sleeping
+            state = :returned_from_yield
+            return rv
+          end
+        end
+      ensure
+        state_lock.synchronize do
+          if state == :sleeping || state == :returned_from_yield
+            state = :killing
+            if y
+              y.kill
+              y.join # make sure y is dead.
+            end
+          end
+        end
+      end
+    rescue exception => e
+      rej = /\A#{Regexp.quote(__FILE__)}:#{__LINE__-4}\z/o
```

Comment:
... and on closer inspection, it seems to have been quite some time since `__LINE__-4` actually pointed where it intended: https://github.com/ruby/ruby/commit/8435c6bb8cd2bedcd95cd5dcc4e5a75ded79735d#diff-32f7163092f0e8e64a3936fba1a3d59dR50

/cc @nobu